### ISSUE TITLE
Pytraj test dependency on cpptraj test cases

### ISSUE THREAD

**Issue by ctlee** [Amber-MD/pytraj#985]
I noticed that the pytraj long tests are dependent upon some of the test cases provided in cpptraj. On my system, I didn't have cpptraj configured in the location expected by the script. Perhaps this should check $CPPTRAJHOME and use that?

```
test_ired_need_lapack_cpptraj (test_ired_TODO.TestIred) ... Error: Topology '../cpptraj/test/Test_IRED/1IEE_A_prot.prmtop' does not exist.
Error: Could not open topology '../cpptraj/test/Test_IRED/1IEE_A_prot.prmtop'
Error: No topologies loaded.
Error: No topology for input trajectory '../cpptraj/test/Test_IRED/1IEE_A_test.mdcrd'
Error: No topologies loaded.
Error: createcrd: No parm files loaded.
Error: Could not initialize action [createcrd]
 ** On entry to DSPEV parameter number  7 had an illegal value
```

**Comment by hainm** [comment]
yeah, we should use CPPTRAJHOME.

I will update this: https://github.com/Amber-MD/pytraj/blob/master/pytraj/testing.py#L12-L21 and the tests to not use relative dir. 

thanks

**Comment by hainm** [comment]
@ctlee this issue should be solved by https://github.com/Amber-MD/pytraj/pull/986

Let me know if you can make it work and if there are other issues.

thanks

**Comment by ctlee** [comment]
Confirmed #986 resolves the issue with running the long test cases. Some notes:
- I'm failing `test_nupars_vs_x3dna`, I'll take a look at what's going on here a bit later.
- The last section Name, Stmts, Miss, Branch etc. I'm not sure I understand! Presumably this is covered in the documentation?

**Comment by hainm** [comment]
> The last section Name, Stmts, Miss, Branch etc. I'm not sure I understand! Presumably this is covered in the documentation?

:D, it's partially covered in doc. You can have a look at developer guide: http://amber-md.github.io/pytraj/doc/build/html/developer_guide.html#code-coverage-how-well-are-your-codes-tested

In summary, I am using `coverage` module to measure how well pytraj's code is coveraged during the tests. I did not understand the meaning of 'Name', 'Stmts', Miss, ... either. You can ignore them.

But you can look at last two words: `Cover   Missing`

`Cover` shows how many percent of code in each file was executed when you are running your tests. 
`Missing` show which lines are not executed. If you want to increase the coverage (ideal 100% of code would be executed), you need to write more tests.

```bash
Name        Stmts   Miss Branch BrPart  Cover   Missing
-----------------------------------------------------------------
pytraj.py     168     27     12      1    80%   240-250, 254-257, 329-334, 341-342
```

Hope this is clear to you.

**Comment by ctlee** [comment]
Oh wow this is great! Very useful measure.

**Comment by hainm** [comment]
yeah, coverage makes your code more reliable since it is well tested (especially if you get 100% coverage)

**Comment by hainm** [comment]
@ctlee what's wrong with 'test_nupars_vs_x3dna' test?

**Comment by ctlee** [comment]
Looks like the arrays `nu.minor`, `nu.major` are different lengths than expected.

**Comment by hainm** [comment]
I see. This is because @drroe made  a PR to cpptraj to add option to match to 3DNA output yesterday.

https://github.com/Amber-MD/cpptraj/pull/152

pytraj used `groovecalc 3dna` as default while cpptraj does not (for backward compat)

https://github.com/Amber-MD/pytraj/blob/master/pytraj/_nastruct.py

**Comment by ctlee** [comment]
Aha I see. Just to confirm function, I noticed some interesting lines in the long unittest output,

```
test_0 (test_drmsd.Test) ... Error: No topologies loaded.
Error: No topologies loaded.
ok
test_vs_cpptraj (test_replicate_cell.TestReplicateCell) ... Error: No topologies loaded.
ok
test_vs_list_tuple (test_replicate_cell.TestReplicateCell) ... Error: No topologies loaded.
Error: No topologies loaded.
ok
```
Looks like the line `pt.distance_rmsd(traj)` outputs this error. Inspecting the trajectory:
```
>>> traj
pytraj.TrajectoryIterator, 101 frames: 
Size: 0.0000 (GB)
<Topology: 223 atoms, 13 residues, 1 mols, non-PBC>
```

**Comment by hainm** [comment]
> Error: No topologies loaded.

this has been a puzzle to me a while. Honestly I don't know why this happens. It's interesting that pytraj can still reproduce cpptraj's output.

- Run completly by cpptraj style

```python
In [8]: txt = '''
parm data/tz2.parm7
trajin data/tz2.nc
drmsd
'''

In [9]: state = pt.load_cpptraj_state(txt)

In [10]: state.run()
Out[10]: 
CpptrajState, include:
<datasetlist: 2 datasets>

In [11]: state.data[-1]
Out[11]: 
<pytraj.datasets.DatasetDouble: size=101, key=DRMSD_00001> 
values: 
[ 0.          2.95729211  4.49201771 ...,  5.186567    5.37325298
  4.82324321]
```

- Run by pytraj's style

```python
In [12]: traj
Out[12]: 
pytraj.TrajectoryIterator, 101 frames: 
Size: 0.0005 (GB)
<Topology: 223 atoms, 13 residues, 1 mols, non-PBC>
           

In [13]: pt.distance_rmsd(traj)
Error: No topologies loaded.
Out[13]: 
array([ 0.        ,  2.95729211,  4.49201771, ...,  5.186567  ,
        5.37325298,  4.82324321])
```

**Comment by hainm** [comment]
@ctlee by the way, I highly recommend to use `ipython` (like my demo above) or `ipython notebook` for interactive work. You can find a live version of pytraj with ipython notebook here. http://app.mybinder.org/3127847493/notebooks/index.ipynb

**Comment by drroe** [comment]
> Error: No topologies loaded. 

@hainm FYI in cpptraj this error message only happens if `DataSetList::GetTopology()` or `GetTopByIndex()` is called and there are no topology data sets in the DataSetList.

**Comment by ctlee** [comment]
@hainm that's a great idea. I'll give it a shot later!

**Comment by hainm** [comment]
@drroe  any idea why this happen to `Action_DistRmsd` only? And why pytraj still get correct result (vs cpptraj)

The code in pytraj is very simple (and pytraj applied the same copy and paste strategy for other actions)

```python
@_register_pmap
@_super_dispatch(has_ref=True)
def calc_distance_rmsd(traj=None, ref=0, mask='', top=None, dtype='ndarray',
        frame_indices=None):
    '''compute distance rmsd between traj and reference

    Parameters
    ----------
    traj : Trajectory-like or iterator that produces Frame
    ref : {int, Frame}, default 0 (1st Frame) 
    mask : str
    top : Topology or str, optional, default None
    dtype : return dtype, default 'ndarray'

    Returns
    -------
    1D ndarray if dtype is 'ndarray' (default)

    Examples
    --------
    >>> import pytraj as pt
    >>> # compute distance_rmsd to last frame
    >>> pt.distance_rmsd(traj, ref=-1)

    >>> # compute distance_rmsd to first frame with mask = '@CA'
    >>> pt.distance_rmsd(traj, ref=0, mask='@CA')
    '''
    dslist = CpptrajDatasetList()
    command = mask
    _ref = _get_reference_from_traj(traj, ref)
    _top = _get_topology(traj, top)
    fi = _get_fiterator(traj, frame_indices)

    act = CpptrajActions.Action_DistRmsd()
    act(command, [_ref, fi], top=_top, dslist=dslist)

    # exclude ref value
    for d in dslist:
        d.data = d.data[1:]
    return _get_data_from_dtype(dslist, dtype=dtype)
```

My wild guess is we need another Topology for reference, right?

**Comment by drroe** [comment]
> My wild guess is we need another Topology for reference, right?

What could potentially cause this is if pytraj does not put topologies in the DataSetList. I think I may need to alter the conditions under which the error message is printed. I'll do some tests and report back.

**Comment by hainm** [comment]
> What could potentially cause this is if pytraj does not put topologies in the DataSetList.

you're right that pytraj currently does not add topology to DSL for this Action.

**Comment by drroe** [comment]
> you're right that pytraj currently does not add topology to DSL for this Action.

Then this is an overzealous error message from cpptraj. I will modify accordingly.

**Comment by drroe** [comment]
Should be fixed by Amber-MD/cpptraj#157

**Comment by hainm** [comment]
confirmed.
